File: src/main/java/net/igneo/icv/enchantmentActions/enchantManagers/armor/helmet/GravityWellManager.java
package net.igneo.icv.enchantmentActions.enchantManagers.armor.helmet;

import net.igneo.icv.client.indicators.EnchantIndicator;
import net.igneo.icv.client.indicators.StasisCooldownIndicator;
import net.igneo.icv.enchantment.EnchantType;
import net.igneo.icv.enchantmentActions.enchantManagers.armor.ArmorEnchantManager;
import net.igneo.icv.init.ICVUtils;
import net.minecraft.core.Direction;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.HitResult;
import net.minecraft.world.phys.Vec3;

import java.util.List;

public class GravityWellManager extends ArmorEnchantManager {
    private Direction surfaceDirection;
    private Vec3 position;
    private static final double PUSH_STRENGTH = 1.5d;
    private static final double EFFECT_RADIUS = 5.0d;
    
    public GravityWellManager(Player player) {
        super(EnchantType.HELMET, 300, 10, false, player);
    }
    
    @Override
    public void onOffCoolDown(Player player) {
        // Method left empty as in original
    }
    
    @Override
    public EnchantIndicator getIndicator() {
        return new StasisCooldownIndicator(this);
    }
    
    @Override
    public void activate() {
        if (!(player.level() instanceof ServerLevel level)) {
            return;
        }
        
        HitResult hitResult = player.pick(5, 0f, false);
        
        if (hitResult.getType() == HitResult.Type.BLOCK) {
            BlockHitResult blockHitResult = (BlockHitResult) hitResult;
            getBlockFacingDirection(blockHitResult);
        }
    }
    
    private void getBlockFacingDirection(BlockHitResult blockHitResult) {
        surfaceDirection = blockHitResult.getDirection();
        position = blockHitResult.getLocation();
        
        applyGravityEffect(surfaceDirection);
    }
    
    private void applyGravityEffect(Direction direction) {
        if (!(player.level() instanceof ServerLevel level)) {
            return;
        }
        
        List<Entity> entities = ICVUtils.collectEntitiesBox(level, position, 7.5f);
        Vec3 pushVector = getPushVector(direction);
        
        for (Entity entity : entities) {
            double distance = entity.position().distanceTo(position);
            double scaledStrength = PUSH_STRENGTH * (1.0 - (distance / EFFECT_RADIUS));
            
            if (scaledStrength <= 0) {
                continue;
            }
            
            Vec3 scaledPush = pushVector.scale(scaledStrength);
            entity.setDeltaMovement(entity.getDeltaMovement().add(scaledPush));
            
            if (entity instanceof LivingEntity && direction == Direction.UP) {
                entity.fallDistance = 0.0F;
            }
        }
    }
    
    private Vec3 getPushVector(Direction direction) {
        return switch (direction) {
            case UP -> new Vec3(0, PUSH_STRENGTH * 0.2f, 0);
            case DOWN -> new Vec3(0, -PUSH_STRENGTH, 0);
            case NORTH -> new Vec3(0, 0.2f, -PUSH_STRENGTH * 2);
            case SOUTH -> new Vec3(0, 0.2f, PUSH_STRENGTH * 2);
            case EAST -> new Vec3(PUSH_STRENGTH * 2, 0.2f, 0);
            case WEST -> new Vec3(-PUSH_STRENGTH * 2, 0.2f, 0);
        };
    }
    
    @Override
    public boolean canUse() {
        return stableCheck();
    }
}

File: src/main/java/net/igneo/icv/enchantmentActions/enchantManagers/armor/helmet/GravityWellManager.kt
package net.igneo.icv.enchantmentActions.enchantManagers.armor.helmet

import net.igneo.icv.client.indicators.EnchantIndicator
import net.igneo.icv.client.indicators.StasisCooldownIndicator
import net.igneo.icv.enchantment.EnchantType
import net.igneo.icv.enchantmentActions.enchantManagers.armor.ArmorEnchantManager
import net.igneo.icv.init.ICVUtils
import net.minecraft.core.Direction
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.phys.BlockHitResult
import net.minecraft.world.phys.Vec3

class GravityWellManager(player: Player?) : ArmorEnchantManager(EnchantType.HELMET, 300, 10, false, player) {
    private var surfaceDirection: Direction? = null
    private var position: Vec3? = null

    override fun onOffCoolDown(player: Player?) = Unit
    override val indicator: EnchantIndicator = StasisCooldownIndicator(this)

    override fun activate() {
        if (player.level() !is ServerLevel) return
        (player.pick(5.0, 0.0f, false) as? BlockHitResult)?.let {
            surfaceDirection = it.direction
            position = it.location
            applyGravityEffect(it.direction)
        }
    }

    private fun applyGravityEffect(direction: Direction) {
        val level = player.level() as? ServerLevel ?: return
        val pos = position ?: return
        val entities = ICVUtils.collectEntitiesBox(level, pos, 7.5)
        val pushVec = when (direction) {
            Direction.UP -> Vec3(0.0, PUSH_STRENGTH * 0.2, 0.0)
            Direction.DOWN -> Vec3(0.0, -PUSH_STRENGTH, 0.0)
            Direction.NORTH -> Vec3(0.0, 0.2, -PUSH_STRENGTH * 2)
            Direction.SOUTH -> Vec3(0.0, 0.2, PUSH_STRENGTH * 2)
            Direction.EAST -> Vec3(PUSH_STRENGTH * 2, 0.2, 0.0)
            Direction.WEST -> Vec3(-PUSH_STRENGTH * 2, 0.2, 0.0)
            else -> Vec3.ZERO
        }

        for (entity in entities) {
            val distance = entity.position().distanceTo(pos)
            val strength = PUSH_STRENGTH * (1 - distance / EFFECT_RADIUS)
            if (strength <= 0) continue

            entity.deltaMovement = entity.deltaMovement.add(pushVec.scale(strength))
            if (entity is LivingEntity && direction == Direction.UP) entity.fallDistance = 0f
        }
    }

    companion object {
        private const val PUSH_STRENGTH = 1.5
        private const val EFFECT_RADIUS = 5.0
    }
}
